### PR TITLE
Isobstructed

### PR DESCRIPTION
**Pull request by erniejamison**
This is close but needs to be looked over. Instead of just returning true or false thethere is a method that returns an array of obstructing pieces. By convention a Ruby method ending with a '?' returns true or false so I separated it and called #obstructing_pieces with it. More tests will be added.

### CODE REVIEW

File: .byebug_history [hunk 1]
```diff
@@ -0,0 +1,50 @@
+c
```
Comment (lasley) [inline .byebug_history]: Should this be checked into source control? Maybe make a PR to add to `.gitignore`?

File: app/models/.byebug_history [hunk 2]
```diff
@@ -0,0 +1 @@
+step
```
Comment (lasley) [inline app/models/.byebug_history]: Should this be checked into source control? Maybe make a PR to add to `.gitignore`?

File: app/models/piece.rb [hunk 3]
```diff
@@ -1,4 +1,39 @@
...
+    elsif moving_diagonally? # needs word
+      (x_start + 1..x_destination - 1).each do |x|
+        y = x - x_start + y_start
+        pieces.concat(Piece.where(x_position: x, y_position: y))
+      end
+    end
+  end
+
+  def is_obstructed?(x_destination, y_destination)
+    obstructing_pieces(x_destination, y_destination)
```
Comment (lasley) [inline app/models/piece.rb]: Why just calling a method? Are you instead trying to [typecast to bool](http://rubyquicktips.com/post/583755021/convert-anything-to-boolean)?

File: coverage/.last_run.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (lasley) [inline coverage/.last_run.json]: Add this directory `coverage/*` to gitignore as part of the PR previously mentioned

File: coverage/.resultset.json [hunk 5]
```diff
@@ -0,0 +1,1065 @@
+{
```
Comment (lasley) [inline coverage/.resultset.json]: Add this directory `coverage/*` to gitignore as part of the PR previously mentioned

File: app/models/piece.rb [hunk 6]
```diff
@@ -1,4 +1,39 @@
...
+    y_start = y_position
+    @x_destination = x_destination
+    @y_destination = y_destination
+    pieces = []
+
+    if moving_horizontally?
+      pieces = Piece.where(game_id: game.id, y_position: y_start, x_position: (x_start + 1..x_destination - 1)).exists?
+    elsif moving_vertically?
+      pieces = Piece.where(game_id: game.id, x_position: x_start, y_position: (y_start + 1..y_destination - 1)).exists?
+    elsif moving_diagonally? # needs word
```
Comment (lasley) [inline app/models/piece.rb]: needs word?

File: spec/models/piece_spec.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
+require 'rails_helper'
+
+RSpec.describe Piece, type: :model do
+  describe 'is_obstructed?' do
+    it 'should detect an obstructing piece horizontally' do
+      game = Game.create
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 3, y_position: 0)
```
Comment (lasley) [inline spec/models/piece_spec.rb]: This create should be in a before or a let statement in order to DRY

File: spec/models/piece_spec.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      Piece.create(game_id: game.id, x_position: 5, y_position: 0)
```
Comment (lasley) [inline spec/models/piece_spec.rb:9]: You should then base this create off of the one that was created centrally

File: spec/models/piece_spec.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      expect(piece.is_obstructed?(7, 0)).to be true
+    end
+
+    it 'should detect an obstructing piece vertically' do
+      game = Game.create
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 2)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 6)
```
Comment (lasley) [inline spec/models/piece_spec.rb]: You should then base this create off of the one that was created centrally

File: spec/models/piece_spec.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 3, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 5, y_position: 0)
+      expect(piece.is_obstructed?(7, 0)).to be true
+    end
+
+    it 'should detect an obstructing piece vertically' do
+      game = Game.create
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 2)
```
Comment (lasley) [inline spec/models/piece_spec.rb]: You should remove this

File: spec/models/piece_spec.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      Piece.create(game_id: game.id, x_position: 0, y_position: 6)
+      expect(piece.is_obstructed?(0, 7)).to be true
+    end
+
+    it 'should detect an obstructing piece diagonally' do
+      game = Game.create
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 4, y_position: 4)
```
Comment (lasley) [inline spec/models/piece_spec.rb]: You should then base this create off of the one that was created centrally

File: spec/models/piece_spec.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 2)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 6)
+      expect(piece.is_obstructed?(0, 7)).to be true
+    end
+
+    it 'should detect an obstructing piece diagonally' do
+      game = Game.create
+      piece = Piece.create(game_id: game.id, x_position: 0, y_position: 0)
+      Piece.create(game_id: game.id, x_position: 0, y_position: 0)
```
Comment (lasley) [inline spec/models/piece_spec.rb]: You should remove this

File: spec/models/piece_spec.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+      Piece.create(game_id: game.id, x_position: 4, y_position: 4)
+      expect(piece.is_obstructed?(7, 7)).to be true
+    end
+  end
```
Comment (lasley) [inline spec/models/piece_spec.rb:57]: Please also add negative test cases - what if I'm not obstructed horizontally, vertically, or diagonally? You'd be surprised how often the negative cases break

File: app/models/piece.rb [hunk 14]
```diff
@@ -1,4 +1,39 @@
 class Piece < ActiveRecord::Base
   belongs_to :game
   belongs_to :player
+
+  def obstructing_pieces(x_destination, y_destination)
```
Comment (lasley) [inline app/models/piece.rb]: I would call this `get_obstructing_pieces`. Typically should follow a `verb_adverb_noun_prounoun` type order

File: app/models/piece.rb [hunk 15]
```diff
@@ -1,4 +1,39 @@
...
+    x_start = x_position
+    y_start = y_position
+    @x_destination = x_destination
+    @y_destination = y_destination
+    pieces = []
+
+    if moving_horizontally?
+      pieces = Piece.where(game_id: game.id, y_position: y_start, x_position: (x_start + 1..x_destination - 1)).exists?
```
Comment (lasley) [inline app/models/piece.rb]: A lot of repetition here, what if you instead
- set default variables for `y_position` (y_start) and `x_position` (x_start) 
- then modify them in the ifs (`_start + 1..x_destination - 1)`)
- then perform one query afterwards on the vars

**Review by lasley — CHANGES_REQUESTED** [review]

File: app/models/piece.rb (hunk 6 above)
Comment (erniejamison) [reply app/models/piece.rb]: work

File: .byebug_history (hunk 1 above)
Comment (erniejamison) [reply .byebug_history]: So I'd have to create a new branch?

Comment (erniejamison) [reply .byebug_history]: Taken care of in another branch

File: app/models/piece.rb (hunk 15 above)
Comment (erniejamison) [reply app/models/piece.rb]: I am stuck on this and need to get help with it.

**Comment by lasley** [comment]
@erniejamison - This issue is marked as Help Wanted, but I don't see any sort of issues detailed here. Is this incorrectly tagged?

**Comment by erniejamison** [comment]
I am going in circles trying to implement the reviews and rewrite the code.

**Comment by erniejamison** [comment]
I am switching this back to returning true/false instead of an array. That is all it needs to do. I'm still working on the diagonal case. Not sure why the files and directories that have been added to .gitignore showed up.

**Comment by erniejamison** [comment]
Please review. So close. Advise on rubocop offenses. Do I need to break up the if/else in the obstructed? method? Will that take care of the 'Assignment Branch Condition size for obstructed? is too high' error?

**Comment by erniejamison** [comment]
Ready for review

File: app/models/piece.rb [hunk 16]
```diff
@@ -1,4 +1,38 @@
...
+    y = y_position
+
+    if moving_horizontally?(x_destination, y_destination)
+      x = (x_position + 1..x_destination - 1)
+    elsif moving_vertically?(x_destination, y_destination)
+      y = (y_position + 1..y_destination - 1)
+    elsif moving_diagonally?(x_destination, y_destination)
+      (x_position + 1..x_destination - 1).each do |x_temp|
+        y = x_temp - x_position + y_position
+        pieces.concat(Piece.where(x_position: x, y_position: y))
```
Comment (erinmorrissey) [inline app/models/piece.rb]: Wouldn't we also want to verify `game.id` here?

File: app/models/piece.rb [hunk 17]
```diff
@@ -1,4 +1,38 @@
...
 class Piece < ActiveRecord::Base
   belongs_to :game
   belongs_to :player
+
+  def obstructed?(x_destination, y_destination)
+    pieces = []
+    x = x_position
+    y = y_position
+
+    if moving_horizontally?(x_destination, y_destination)
```
Comment (erinmorrissey) [inline app/models/piece.rb]: Is there a reason for passing in `x_destination` when only `y_destination` is needed for the method?

Comment (erniejamison) [reply app/models/piece.rb]: Not sure what you mean. The method uses x_destination. If I remove it I get wrong number of argument errors. Even though x and y destination are not being used in the helper methods the method still expects 2 arguments. Can you elaborate?

File: app/models/piece.rb (hunk 16 above)
Comment (erniejamison) [reply app/models/piece.rb]: Probably I'll look at it

File: app/models/piece.rb (hunk 17 above)
Comment (erinmorrissey) [reply app/models/piece.rb]: Inside of the `moving_horizontally?` method, only the `y_destination` argument is used. Just seems cleaner to change `moving_horizontally?` to accept only 1 argument and change the `if moving_horizontally?` method call to pass only 1 argument, but I wasn't sure if I was missing something. Same would go for the `moving_vertically?` method...

**Comment by erniejamison** [comment]
Ready for review

File: app/models/piece.rb (hunk 17 above)
Comment (erniejamison) [reply app/models/piece.rb]: I got you. Makes sense. Thanks

File: .byebug_history (hunk 1 above)
Comment (lasley) [reply .byebug_history]: This file still needs to be removed from this diff

File: app/models/.byebug_history (hunk 2 above)
Comment (lasley) [inline app/models/.byebug_history]: This file should be removed

**Review by lasley — CHANGES_REQUESTED** [review]
Minor fixes, just file deletions. Everything else 👍

**Review by lasley — APPROVED** [review]

**Comment by lasley** [comment]
Thanks @erniejamison

**Comment by erinmorrissey** [comment]
👍  merging!

**Pull request merged**